src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -621,16 +711,42 @@ public bool HasSkipLocalsInitAttribute
         }
 
         internal bool IsAutoPropertyWithGetAccessor
-            => IsAutoProperty && _getMethod is object;
+            => IsAutoProperty && GetMethod is object;
 
-        protected bool IsAutoProperty
-            => (_propertyFlags & Flags.IsAutoProperty) != 0;
+        internal bool IsAutoProperty
+        {
+            get
+            {
+                EnsureAccessorsBinding();
```

Comment:
>EnsureAccessorsBinding(); [](http://example.com/codeflow?start=16&length=25)

Why is it necessary to do this? Why do we want this property to reflect presence of "field" keyword? Where do we depend on that?

Reply:
@AlekseyTs One of the important places is `MethodBodySynthesizer.ConstructAutoPropertyAccessorBody`

Reply:
Actually I think I'm wrong. This has nothing to do with `ConstructAutoPropertyAccessorBody`. Will need to re-check things.

Reply:
@AlekseyTs We depend on that in `GetFieldsToEmit`:

```csharp
                        if (m is SourcePropertySymbol propertySymbol)
                        {
                            FieldSymbol? backingField = propertySymbol.BackingField;
                            if (backingField is SynthesizedBackingFieldSymbol { IsCreatedForFieldKeyword: true })
                            {
                                Debug.Assert(!members.Contains(backingField));
                                yield return backingField;
                            }
                        }
```

Reply:
> We depend on that in GetFieldsToEmit

I do not see ```GetFieldsToEmit``` referencing ```IsAutoProperty```. What am I missing? The question remains, why do we need to call ```EnsureAccessorsBinding``` in ```IsAutoProperty```?

Reply:
For some reason I thought the comment is on `BackingField`, not `IsAutoProperty`.

Binding is required for `IsAutoProperty` too. The binding will not only create the backing field, but also change the auto prop flag.

The property with initializer test should fail if we removed the binding because of:

https://github.com/dotnet/roslyn/blob/8ef33d043aab2d17f675696055d814cb49bb81a1/src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs#L674

Reply:
> The binding will not only create the backing field, but also change the auto prop flag.

I am questioning this behavior too.

>The property with initializer test should fail if we removed the binding because of:

[https://github.com/dotnet/roslyn/blob/8ef33d043aab2d17f675696055d814cb49bb81a1/src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs#L674](https://github.com/dotnet/roslyn/blob/8ef33d043aab2d17f675696055d814cb49bb81a1/src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs#L674)

For me this just indicates that we are doing things too early. If we are not ready to report diagnostics at that place because we do not know yet whether there is a synthesized backing field, we should look for a better place to perform the check. Here is where I would like us to get - in a normal compilation scenario, without any SemanticModel requests, we should not bind accessors multiple times.

Reply:
@AlekseyTs

> because we do not know yet whether there is a synthesized backing field

We *already* know there is a synthesized backing field, we create the field in `SourcePropertySymbolBase` constructor:

https://github.com/Youssef1313/roslyn/blob/c40f61c9e86dfb61b06d94af5a049a074994c678/src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs#L168-L172

But we haven't yet determined whether it's an auto property.

Reply:
> We already know there is a synthesized backing field, we create the field in 

>But we haven't yet determined whether it's an auto property.

Specifics are not important. Whatever you think we haven't yet figured out and therefore need to bind the body. I suggest starting with the original behavior of this property. Without binding accessors and without adjusting flags during binding. If that behavior of the property doesn't match your new definition of what an auto property is, rename it so that it no longer confuses you. After that I would see what doesn't work as we want it to work. Then I would understand why it doesn't work, then I would figure out what is the right way to make it work (not the easiest way, but the right way). I can help you with that. If some issues are too tricky to deal with and you don't want to spend too much time in context of this PR. I would prefer us to disable failing tests with prototype comments and following up on those issues separately, rather than keeping the tests passing with implementation that we don't like.

Reply:
> I have tried to remove the auto prop flag and just rely on the existence of the backing field, I don't remember why this didn't go smooth though.

There is a possibility that we will arrive back to the current implementation, but we should have a clear understanding why this implementation is absolutely  necessary (the key word is "absolutely", i.e. there is no other viable alternative). And we should be able to clearly articulate that.

Reply:
> If that behavior of the property doesn't match your new definition of what an auto property is, rename it so that it no longer confuses you.

A new name would then be "IsAutoPropertyWithASemicolonOnlyAccessor" or "HasASemicolonOnlyAccessor", or whatever good name with the same meaning.

---

> After that I would see what doesn't work as we want it to work.

The thing that shouldn't work given the new name is the `CheckInitializer(IsAutoProperty, ....)` line

I see we can change that to `IsAutoPropertyWithSemicolonOnlyAccessor || BackingField is { IsCreatedForFieldKeyword: true }`, this is also related to your other feedback about "IsCreatedForFieldKeyword" have an incorrect value, which is something I had missed.

Do you think  `IsAutoPropertyWithSemicolonOnlyAccessor || BackingField is { IsCreatedForFieldKeyword: true }` is the right way to do it?

Reply:
Hmmm, actually I can leave the name as "IsAutoProperty" and add the `|| BackingField is { IsCreatedFieldKeyword: true }`, which is obviously much simpler than doing binding.

Reply:
> Actually BackingField getter will do binding.

Exactly. I would like us to be very cautious about triggering binding of accessors. Finding ways to avoid that
and, if there is no other way, be vary explicit that this is what is going to happen. Let me finish reviewing what you have up to date (up to commit 40) and then I'll try to make a recommendation how to move this commit forward, i.e. to get it to the state that can be merged into a feature branch.